### DISCUSSION TITLE
Guidelines to port Bao to a new board

### DISCUSSION THREAD

**Discussion by robertoxmed** [bao-project/bao-hypervisor#54] [General]
Hello,

I have been (very) slowly working on a port of Bao to a new board based on the Rockchip rk3399 (4 x CortexA53 + 2 x Cortex A72). The 2 boards I have with this CPU support U-Boot and ATF by default.

I am working from the demo repository trying to get the baremetal demo to work. Would like to avoid swapping the U-boot and ATF that are installed on my board and just send the new `bao.img` to the board.

Is there a better way to go about this? I see there is a clone of ATF for Bao but I am not very familiar with this component, so maybe I am missing something.

**Comment by josecm** [comment]
Hey @robertoxmed, 

The only reason we have an ATF fork is that in some boards (e.g. imx8qm) we need to configure the stream ids of some devices according to the VM they are assigned to, so we can later program the SMMU using those ids. In some platforms (e.g. zcu104), these ids are hardwired so for such boards we can also use stock ATF. If you don't need to do this kind of thing for that board, I think you can use any ATF  you'd like (provided it provides psci >= 0.2 which is assumed by bao). 

There are also no restrictions regarding U-boot. Technically, in some platforms, we could actually boot bao directly from ATF without going through uboot.

You can also go without passing bao's binary through mkimage. In uboot you can just directly load the raw bao.bin to memory (needs to be a 2MiB aligned address) and run `go <load-address>`.

In conclusion, your approach should work. I don't see any reason it shouldn't.
